src/main/java/no/digipost/stream/NonEmptyStream.java

```diff
@@ -0,0 +1,401 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package no.digipost.stream;
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.Spliterator;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.BinaryOperator;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.function.ToDoubleFunction;
+import java.util.function.ToIntFunction;
+import java.util.function.ToLongFunction;
+import java.util.stream.Collector;
+import java.util.stream.DoubleStream;
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+import static no.digipost.DiggBase.friendlyName;
+
+/**
+ * A stream which is guarantied to produce at least one element. It provides
+ * some extensions to the general {@link Stream} API for processing the stream knowing
+ * it is non-empty, including, but not limited to:
+ * <ul>
+ *   <li>{@link #first()} and {@link #any()}</li>
+ *   <li>{@link #limitToNonEmpty(long)}</li>
+ *   <li>{@link #collect(EmptyIfEmptySourceCollector)}</li>
+ *   <li>{@link #reduceFromFirst(BinaryOperator)}</li>
+ * </ul>
+ *
+ * A selection of operations which maintains the cardinality of the stream returns {@code NonEmptyStream},
+ * such as:
+ * <ul>
+ *   <li>{@link #map(Function)}
+ * </ul>
+ * This selection may be extended in the future. For operations which can not guarantie the
+ * result will still be a non-empty stream, such as {@code filter} and {@code limit}, will yield a
+ * regular {@link Stream}.
+ *
+ *
+ * @param <T> the type of the stream elements
+ */
+public class NonEmptyStream<T> implements Stream<T> {
+
+    public static <T> NonEmptyStream<T> of(T singleElement) {
+        return of(singleElement, Stream.empty());
+    }
+
+    @SafeVarargs
+    public static <T> NonEmptyStream<T> of(T firstElement, T ... remainingElements) {
+        return of(firstElement, Arrays.stream(remainingElements));
+    }
+
+    public static <T> NonEmptyStream<T> of(T firstElement, Stream<T> remainingElements) {
+        return of((Supplier<T>) () -> firstElement, remainingElements);
+    }
+
+    public static <T> NonEmptyStream<T> of(Supplier<? extends T> firstElement, Stream<T> remainingElements) {
+        return new NonEmptyStream<>(firstElement, remainingElements);
+    }
+
+
+    private final Stream<T> completeStream;
+
+    private NonEmptyStream(Supplier<? extends T> firstElement, Stream<T> remainingElements) {
+        this(Stream.concat(Stream.generate(firstElement).limit(1), remainingElements));
```

Comment:
Hva gjør `limit(1)` her? Vet vi ikke allerede fra `Supplier<? extends T> firstElement` at det bare er ett element?

Reply:
Jo, en Supplier gir ett element, eller mer riktig, den gir ett element for hvert kall på Supplier-en, altså hvert gang noen gjør `Supplier.get()`. Men det er ikke noe i veien for at en Supplier har intern tilstand/sideeffekter som gjør at den kan endre hva som returnere for hvert kall, og`Stream.generate(Supplier)` vil i utgangspunktet bare fortsette å kalle på supplier-en den får med mindre man sier noe annet.

F.eks. vil `Stream.generate(() -> "a")` gi en evig stream med masse "a"-elementer.

`limit(1)` er på stream-en, ikke på supplieren, og sørger for at vi bare er interessert i ett element fra hva enn den får fra `Supplier.get()`.